proposals/factory_constructors.md

```diff
@@ -0,0 +1,216 @@
+yet been observed (outside of whatever a user does in their constructor). This method must be `virtual`, as the value returned should be of the same
+runtime time as the value the `with` expression was called on, which may be more derived than the compile-time type of the expression. However, there
+are some more general motivating patterns that .NET developers use today which can help inform design decisions here:
+
+* Factories: Not factories as defined in this proposal, but the Factory pattern as implemented by many, many DI frameworks. Users use factories today
+in order to decouple their implementations from their interfaces for many different reasons, including testability, ability to swap strategies as
+needed, as a public extension point, and more. These factories tend to consist of anything from a series of static methods in simple cases to full
+classes where the factory itself is injected with many instance methods for constructing things.
+* Static `Create` methods: These methods are often introduced as a substitute for generic inference in constructors. In C# today, there is no generic
+inference for constructor type parameters: they must be fully spelled out. There are a few different proposals for partial inference or full inference
+today, but none have been implemented.
+
+This proposal does not attempt to solve the second bullet here, but is intended to be delivered at the same time a solution to that problem as a
+general construction improvement iteration to C#.
+
+## Detailed Design
+
+We introduce a new type of constructor: a factory constructor. These constructors do not have to be defined in the same class that they create, can
+be virtual, can be static, and can return more derived instances of the type they construct. These factory constructors have restrictions on what they
+can return, much like the v1 proposal has, in order to ensure that the instance returned is unobservered outside of the factory call.
+
+```cs
+public class Widget
+{
+    public string Name { get; set; }
+    public int Id { get; init; }
+
+    internal Widget() { }
+}
+
+public static class WidgetFactory
+{
+    public static factory Widget() => new Widget();
+}
+
+var w = new WidgetFactory.Widget()
+{
+    Name = "333fred",
+    Id = 43
+};
+```
+
+### Grammar
+
+Factory constructors are specified by the following grammar:
+
+```antlr
+factory_constructor_declaration
+    : attributes? factory_constructor_modifier* 'factory' factory_constructor_declarator factory_constructor_body
+    ;
+
+factory_constructor_modifier
+    : 'new'
+    | 'public'
+    | 'protected'
+    | 'internal'
+    | 'private'
+    | 'static'
+    | 'virtual'
+    | 'sealed'
+    | 'override'
+    | 'abstract'
+    | factory_constructor_modifier_unsafe
+    ;
+
+factory_constructor_modifier_unsafe
+    : 'unsafe'
+    ;
+
+factory_constructor_declarator
+    : type '(' formal_parameter_list? ')'
+    ;
+
+factory_constructor_body
+    : block
+    | ';'
```

Comment:
or `=>`?

Reply:
Yeah. I copied the grammar from the existing spec, which I assume needs to be updated for expression-bodied things.